core/services/workflows/secrets/worker.go

```diff
@@ -0,0 +1,147 @@
+package syncer
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	"github.com/smartcontractkit/chainlink-common/pkg/types"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/workflow/generated/workflow_registry_wrapper"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/utils/chans"
+	"github.com/smartcontractkit/chainlink/v2/core/utils/signalers"
+)
+
+var (
+	defaultTickInterval = 12 * time.Second
+	ContractName        = "WorkflowRegistry"
+	ContractEventName   = "WorkflowForceUpdateSecretsRequestedV1"
+)
+
+type FetcherFunc func(ctx context.Context, url string) ([]byte, error)
+
+type worker struct {
+	services.StateMachine
+	stopCh   services.StopChan
+	eventsCh <-chan workflow_registry_wrapper.WorkflowRegistryWorkflowForceUpdateSecretsRequestedV1
+	timer    <-chan struct{}
+	lggr     logger.Logger
+	cr       types.ContractReader
+	cfg      ContractEventPollerConfig
+	addr     string
+	orm      ORM
+	gateway  FetcherFunc
+	wg       sync.WaitGroup
+}
+
+func WithTimer(t <-chan struct{}) func(*worker) {
+	return func(w *worker) {
+		w.timer = t
+	}
+}
+
+func NewWorker(
+	lggr logger.Logger,
+	startBlockNum uint64,
+	qryCount uint64,
+	contractAddr string,
+	cr types.ContractReader,
+	orm ORM,
+	gateway FetcherFunc,
+	opts ...func(*worker),
+) *worker {
+	w := &worker{
+		lggr:    lggr,
+		cr:      cr,
+		orm:     orm,
+		gateway: gateway,
+		cfg: ContractEventPollerConfig{
+			ContractName:      ContractName,
+			ContractEventName: ContractEventName,
+			ContractAddress:   contractAddr,
+			QueryCount:        qryCount,
+			StartBlockNum:     startBlockNum,
+		},
+		stopCh: make(services.StopChan),
+	}
+
+	if w.cfg.QueryCount == 0 {
+		w.cfg.QueryCount = 20
+	}
+
+	for _, opt := range opts {
+		opt(w)
+	}
+
+	return w
+}
+
+func (w *worker) GetLogs() <-chan workflow_registry_wrapper.WorkflowRegistryWorkflowForceUpdateSecretsRequestedV1 {
+	return w.eventsCh
+}
+
+func (w *worker) Start(_ context.Context) error {
+	return w.StartOnce("secrets_worker", func() error {
+		ctx, cancel := w.stopCh.NewCtx()
+
+		done, _ := w.syncForceUpdateSecretsEvents(ctx, w.cfg)
+
+		w.wg.Add(1)
+		go func() {
+			defer w.wg.Done()
+			defer cancel()
+			<-done
+		}()
+
+		return nil
+	})
+}
+
+func (w *worker) Close() error {
+	return w.StopOnce("secrets_worker", func() error {
+		close(w.stopCh)
+		w.wg.Wait()
+		return nil
+	})
+}
+
+func (w *worker) getTicker(ctx context.Context) <-chan struct{} {
+	if w.timer == nil {
+		w.timer = signalers.MakeTicker(ctx.Done(), defaultTickInterval)
+	}
+	return w.timer
+}
+
+// syncForceUpdateSecretsEvents synchronizes the force update secrets events from the contract
+// to the local database.
+func (w *worker) syncForceUpdateSecretsEvents(
```

Comment:
@MStreet3 YAGNI the concurrency here: it's abstracting things away thus making things a bit harder to understand and I'm not sure what we're gaining here.

This operation should correspond to:
- updating the relevant secrets entry in the DB
- updating the local in-mem cache for the secrets

I'd also recommend combining this package with the syncer package for more cohesion.

Reply:
Hm, there's no extra concurrency here. This method just starts the secrets event query worker and passes the events to the event handler.